### DISCUSSION TITLE
How would we get Cache Tags using GraphQL

### DISCUSSION THREAD

**Discussion by Hyra** [ostark/upper#64] [Ideas]
At the moment i'm working on a headless implementation of Craft with GraphQL

We're used to Upper generating the correct Cache Tags in the header so we can clear by Tag when entries are saved.

However, when using graphql we obviously don't get this information.
I was contemplating on building the cache tags based on what we query (e.g. query the ID for elements as well) but this seems something flaky.

Anyone with a suggestion or hint on how to go about this?

**Comment by ostark** [comment]
Do you use GraphQL with GET requests and query params or POST?
With POST it's probably not suited.

**Reply by Hyra** [reply]
GET requests.

I've got sort of a solution working where I hook into the `EVENT_BEFORE_EXECUTE_GQL_QUERY` and `EVENT_AFTER_EXECUTE_GQL_QUERY` events from GQL and use the `$elementsService->stopCollectingCacheTags();` to retrieve the affected elements and attach that to the output of the query in a property `cacheTags`

Pretty much as lined out here (regarding recording the elements queried): https://github.com/craftcms/cms/issues/1507#issuecomment-633147835

I then transform them with a regular expression to use the `seX` and `elY` syntax so I can use them as headers.

Upper can then do its work as it usually does.

Not yet fully tested in the wild, but at first glance this seems a workable approach.
